src/nvim/buffer.c

```diff
@@ -2855,98 +2849,152 @@ build_stl_str_hl (
+    }
 
-    /*
-     * Handle up to the next '%' or the end.
-     */
-    while (*s != NUL && *s != '%' && p + 1 < out + outlen)
-      *p++ = *s++;
-    if (*s == NUL || p + 1 >= out + outlen)
+    // Copy the formatting verbatim until we reach the end of the string
+    // or find a formatting item (denoted by `%`)
+    // or run out of room in our output buffer.
+    while (*fmt_p != NUL && *fmt_p != '%' && out_p < out_end_p)
+      *out_p++ = *fmt_p++;
+
+    // If we have processed the entire format string or run out of
+    // room in our output buffer, exit the loop.
+    if (*fmt_p == NUL || out_p >= out_end_p)
       break;
 
-    /*
-     * Handle one '%' item.
-     */
-    s++;
-    if (*s == NUL)      /* ignore trailing % */
+    // The rest of this loop wil handle a single `%` item.
+    // Note: We increment here to skip over the `%` character we are currently
+    //       on so we can process the item's contents.
+    fmt_p++;
+
+    // Ignore `%` at the end of the format string
+    if (*fmt_p == NUL) {
       break;
-    if (*s == '%') {
-      if (p + 1 >= out + outlen)
+    }
+
+    // Two `%` in a row is the escape sequence to print a
+    // single `%` in the output buffer.
+    if (*fmt_p == '%') {
+      // Ignore the character if we're out of room in the output buffer.
+      if (out_p >= out_end_p)
         break;
-      *p++ = *s++;
-      prevchar_isflag = prevchar_isitem = FALSE;
+      *out_p++ = *fmt_p++;
+      prevchar_isflag = prevchar_isitem = false;
       continue;
     }
-    if (*s == STL_MIDDLEMARK) {
-      s++;
-      if (groupdepth > 0)
+
+    // STL_MIDDLEMARK denotes the separation place between
+    // left and right aligned items.
+    if (*fmt_p == STL_MIDDLEMARK) {
+      fmt_p++;
+      // Ignored when we are inside of a grouping
+      if (groupdepth > 0) {
         continue;
+      }
       item[curitem].type = Middle;
-      item[curitem++].start = p;
+      item[curitem++].start = out_p;
       continue;
     }
-    if (*s == STL_TRUNCMARK) {
-      s++;
+
+    // STL_TRUNCMARK denotes where to begin truncating if the
+    // statusline is too long.
+    if (*fmt_p == STL_TRUNCMARK) {
+      fmt_p++;
       item[curitem].type = Trunc;
-      item[curitem++].start = p;
+      item[curitem++].start = out_p;
       continue;
     }
-    if (*s == ')') {
-      s++;
-      if (groupdepth < 1)
+
+    // The end of a grouping
+    if (*fmt_p == ')') {
+      fmt_p++;
+      // Ignore if we are not actually inside a group currently
+      if (groupdepth < 1) {
         continue;
+      }
       groupdepth--;
 
-      t = item[groupitem[groupdepth]].start;
-      *p = NUL;
-      l = vim_strsize(t);
+      // { Determine how long the group is.
+      //   Note: We set the current output position to null
+      //         so `vim_strsize` will work.
+      char_u *t = item[groupitem[groupdepth]].start;
+      *out_p = NUL;
+      long l = vim_strsize(t);
+      // }
```

Comment:
Avoid "comment scopes" in most cases. E.g., this comment scope is already fairly clear because of the blank line below. In other cases, a comment on a `while()` is already "scoped" by the `while(){...}` scope.